### PR TITLE
Project update

### PR DESCRIPTION
**Pull request by sean-m-sullivan**
##### SUMMARY
Adding a module to the awx collection that updates project similar to tower_job_launch

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
Collection

##### AWX VERSION
awx: 14.0.0

### CODE REVIEW

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (AlanCoding) [inline awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: maybe I missed something, but doesn't a project with this name need to be created first?

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_project_update.py]: @john-westcott-iv I think it would be interesting to look at shared utilities between this as tower_job_launch, these patterns feel very familiar

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [inline awx_collection/plugins/modules/tower_project_update.py]: I'm torn on this return. We do a similar dict return in tower_workflow_launch. However, in tower_job_cancel, tower_job_launch, and tower_job_wait we are returning specific items (i.e. id, time elapsed, status, etc). I think I like returning specific documented things better than a generic hash with no definition of what is in it. Thoughts?

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [inline awx_collection/plugins/modules/tower_project_update.py:39]: If we are going to make wait optional do we need a second tower_project_wait (like tower_job_launch and tower_job_wait) so that you can launch a project update, do some stuff in the background and then bring the wait into the foreground to get the results? This is really the async function in Ansible but I think our modules and pattern was developer before async was an actual thing.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_project_update.py]: Yes, I agree, we could probably make a wait_until_finished method which took a URL, interval, error msg, etc and performed this loop.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_project_update.py]: It looks like below we are returning specific pieces of information, not a project_info dict.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [inline awx_collection/plugins/modules/tower_project_update.py]: We use both the Sync and Update terms in the module and its name. We should pick one.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [inline awx_collection/plugins/modules/tower_project_update.py]: You can remove required=False here.

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (john-westcott-iv) [inline awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: The object name structure is also: `AWX-Collection-tests-<module>-<asset type>-<id>`. So tower_job_template should be later to tower_project_update. And, since you use it down below it should probably be in a variable. Check out tower_export for an example of creating a test_id, creating multiple assets and then running things in a block to try and ensure cleanup on a failed test.

**Review by john-westcott-iv — CHANGES_REQUESTED** [review]

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (sean-m-sullivan) [reply awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: It does need to be created, which is why I took the info from the tower project job thinking it would be run after.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: I took the tower_job_launch as a template to use to make this, and changed it to update project. So it is definitely the same code.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: Will update that.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py:39]: Good question, I had more thought of it, that You can wait for the results, but a given Job template is blocked until the corresponding project is updated, so if you launch a template afterwards it is still effectively doing the wait. Leaving the option up to what the user decides.

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (beeankha) [inline awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: Any `False` or `True` values in the test playbooks need to be lower case to pass lint check

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (sean-m-sullivan) [reply awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: Updating this to include its own project.

Comment (sean-m-sullivan) [reply awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: Thanks for the tip using that as a reference

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: Updated to include better information about what is returned

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: Do we want to do this now, or do this as a separate update.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_project_update.py:28]: @john-westcott-iv this wouldn't allow specifying the project by id. Do we have a pattern for that?

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_project_update.py]: I'm fine to punt, maybe leave a comment about it

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_project_update.py:39]: I'm always in favor of fewer modules. Maybe we could add the wait option there, and draft an example with `awx.awx.tower_api` lookup plugin to deprecate the tower_job_wait module?

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py:28]: I did not think we have a pattern for that in other modules, but I'll look

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py:39]: If we were to put it in tower_api, we could build it into the "Create" modules like tower_project, tower_job_template, with less required options, that would fail on create, but would allow for an "update", First thing I tried with tower_project was doing this, and it overwrote the things to blank that I didn't provide.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: I'm inclined to punt it atm, but put it on my docket of things to do. Been putting in the work on the schema PR, got a good set of test cases for it, updated Documentation, but ran into snag where its throwing an error, figuring that out.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_project_update.py:28]: In this case it would not because get_one does not attempt to do an ID if it can't find something by name. resolve_name_to_id does allow for a ID and will try an ID after failing to find something by name. But resolve_name_to_id does not allow for additional search criteria. We should think about combining these functionalities.

**Review by john-westcott-iv — COMMENTED** [review]
One kind of high level thing here. For tower_job we have: launch, wait, cancel and list. List may not be applicable to this but do we want to follow the same patter with launch, cancel and wait for tower_project_update? (i.e. make this tower_project_update_launch and add a wait and cancel modules).

File: awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml
Comment (squidboylan) [inline awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: If this returns the job data we should assert that it's done (to verify that `wait` actually works), im not exactly sure what data we get back here

Comment (sean-m-sullivan) [reply awx_collection/tests/integration/targets/tower_project_update/tasks/main.yml]: I've updated this to successful here was the result info
```
{
    "result": {
        "changed": true,
        "failed": false,
        "id": 925,
        "status": "successful",
    }
}
```

**Comment by sean-m-sullivan** [comment]
> One kind of high level thing here. For tower_job we have: launch, wait, cancel and list. List may not be applicable to this but do we want to follow the same patter with launch, cancel and wait for tower_project_update? (i.e. make this tower_project_update_launch and add a wait and cancel modules).
@john-westcott-iv 
So I've looked into this, and at least I thought I'd break down thoughts on these. 

wait: 
because projects are short and don't make changes, I think its applicable to put not have a separate module, especially as tower_project and tower_project update have wait built in. That being said, tower_workflow_launch has wait built in, while tower_job_launch does not, thus requiring job_launch + job wait to achieve the same results. 

cancel:
I believe this is applicable for jobs, but general project updates will either quickly fail, timeout, or succeed and don't make changes versus workflows and jobs where you might want to cancel them via module. 

Launch:
I think this is just down to nomenclature, projects update, jobs launch, as per the api, 
/api/v2/projects/{id}/update/
/api/v2/job_templates/{id}/launch/
/api/v2/workflow_job_templates/{id}/launch/

**Comment by sean-m-sullivan** [comment]
@AlanCoding @john-westcott-iv 

I'm working on the name_to_id, but think it may be beyond the scope of this merge, same with the wait as a function, Do we want to merge this without those items? or should we wait on those changes.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py:28]: This module now resolves by id or name. I tried to do it as a global change but when coupled with project_update, it would change the name to the ID#, as that project hardcodes where it gets the name from the module args. If we want to implement this collection wide, we would need to rewrite all the modules where this is the case.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: Added #8022 to address this, once we have some feedback on this, or even after this is merged, will update this module to use the new method.

**Comment by AlanCoding** [comment]
This is looking good to me, if integration tests pass then I agree it makes sense and then later return to the consolidation of code you've already started on.

**Review by AlanCoding — APPROVED** [review]
I think all the feedback has been addressed. After I get integration results I'll check back in.

**Review by beeankha — APPROVED** [review]

**Comment by sean-m-sullivan** [comment]
recheck

**Comment by sean-m-sullivan** [comment]
@john-westcott-iv wait changed merged, should satisfy the change request.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_project_update.py]: this is breaking tests on python2 because this syntax only works for python3

**Comment by AlanCoding** [comment]
recheck

File: awx_collection/plugins/modules/tower_project_update.py
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project_update.py]: Updated to use isdigit instead, added more tests and better fail messages.

File: awx_collection/plugins/modules/tower_project_update.py
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_project_update.py]: this hit

```
plugins/modules/tower_project_update.py:109:85: W291: trailing whitespace
```

**Comment by sean-m-sullivan** [comment]
recheck

**Review by AlanCoding — APPROVED** [review]

**Comment by sean-m-sullivan** [comment]
regate

**Comment by sean-m-sullivan** [comment]
recheck

**Review by john-westcott-iv — APPROVED** [review]

**Pull request merged**
